tm2/pkg/sdk/bank/keeper.go

```diff
@@ -21,24 +22,84 @@ type BankKeeperI interface {
 	SubtractCoins(ctx sdk.Context, addr crypto.Address, amt std.Coins) (std.Coins, error)
 	AddCoins(ctx sdk.Context, addr crypto.Address, amt std.Coins) (std.Coins, error)
 	SetCoins(ctx sdk.Context, addr crypto.Address, amt std.Coins) error
+
+	InitGenesis(ctx sdk.Context, data GenesisState)
+	GetParams(ctx sdk.Context) Params
 }
 
-var _ BankKeeperI = BankKeeper{}
+var _ BankKeeperI = &BankKeeper{}
 
 // BankKeeper only allows transfers between accounts without the possibility of
 // creating coins. It implements the BankKeeperI interface.
 type BankKeeper struct {
 	ViewKeeper
 
 	acck auth.AccountKeeper
+	// The keeper used to store parameters
+	paramk           params.ParamsKeeper
+	params           Params
+	restrictedDenoms map[string]struct{}
 }
 
 // NewBankKeeper returns a new BankKeeper.
-func NewBankKeeper(acck auth.AccountKeeper) BankKeeper {
-	return BankKeeper{
+func NewBankKeeper(acck auth.AccountKeeper, pk params.ParamsKeeper) *BankKeeper {
+	rdm := map[string]struct{}{}
+
+	params := DefaultParams()
+	for _, denom := range params.RestrictedDenoms {
+		rdm[denom] = struct{}{}
+	}
+	return &BankKeeper{
 		ViewKeeper: NewViewKeeper(acck),
 		acck:       acck,
+		paramk:     pk,
+		params:     params,
+		// Store restricted denoms in a map's keys for fast
+		// comparison when filtering out restricted denoms from a send message.
+		restrictedDenoms: rdm,
+	}
+}
+
+func (bank *BankKeeper) AddRestrictedDenoms(ctx sdk.Context, restrictedDenoms ...string) {
```

Comment:
Can these be called concurrently?
I'm assuming not, since we tightly control access to the `BankKeeper`

Reply:
No concurrent calls are allowed. It can only be accessed from the VM GovDAO.